### ISSUE TITLE
PredisCache "invalid bulk length" error

### ISSUE THREAD

**Issue by jstruzik** [doctrine/cache#210]
Predis version: `1.1.1`
Redis server version: `3.0.7`

I keep getting the error from the Redis server when attempting to retrieve a cached value:

> ERR Protocol error: invalid bulk length

---

What the data looks like after hydrating and passing to the `CacheProvider`:

```
array(1) {
  ["SELECT t0_.id AS id_0 FROM test t0_ WHERE t0_.id = ?-a:1:{i:0;i:1;}-a:1:{i:0;s:7:"integer";}"]=>
  array(1) {
    [0]=>
    array(1) {
      ["id_0"]=>
      string(1) "1"
    }
  }
}
```

Cache setup

```php
$config = Setup::createConfiguration();
$redis_cache = new PredisCache(new PredisClient('connection_string_here'));
$config->setResultCacheImpl($redis_cache);
$entity_manager = EntityManager::create($db_params, $config);
```

TestEntity
```php
class TestEntity
{
    /**
     * @var int
     */
    private $id;

    /**
     * Gets the id.
     *
     * @return int
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * Sets the id.
     *
     * @param int $id
     * @return $this
     */
    public function setId($id)
    {
        $this->id = $id;

        return $this;
    }
}
```

Test Entity Mapping
```php
$builder = new ClassMetadataBuilder($metadata);
$builder->setTable('test');

$builder
    ->createField('id', Type::INTEGER)
    ->nullable(false)
    ->unique(true)
    ->option('unsigned', true)
    ->makePrimaryKey()
    ->build();

$metadata->setIdGenerator(new AssignedGenerator());
$metadata->setIdGeneratorType(ClassMetadata::GENERATOR_TYPE_AUTO);

$metadata->markReadOnly();
```

Example fetch
```php
$test = $entity_manager->createQueryBuilder()
    ->select('test')
    ->from(TestEntity::class, 'test')
    ->where('test.id = :id')
    ->setParameter('id', 1, Type::INTEGER)
    ->getQuery()
    ->setResultCacheId('test_id')
    ->getSingleResult();
```

---

My best guess is that calling `PredisClient::set($key, $data)` where `$data` is a nested array is causing an issue since the data isn't serialized. Or I've just somehow messed up my setup/implementation.

Let me know if there's any more information I can provide!

**Comment by Ocramius** [comment]
@jstruzik this needs to be tested in isolation within the test suite of `doctrine/cache`, not with a mix of cache+ORM

**Comment by jstruzik** [comment]
> @jstruzik this needs to be tested in isolation within the test suite of doctrine/cache, not with a mix of cache+ORM

Ended up writing a unit test to attempt to simulate this behavior with the data:

```php
'entity_array' => [
    'SELECT t0_.id AS id_0 FROM test t0_ WHERE t0_.id = ?-a:1:{i:0;i:1;}-a:1:{i:0;s:7:"integer";}'
        => [
            'id_0' => '1'
    ]
]
```

Which ran without a problem of course. After a little digging, I realized that my project was locked on version `1.4.1`... This issue has seemed to be fixed with proper serialization in https://github.com/doctrine/cache/pull/78 and merged in `1.4.2`. I have since updated and no longer get the issue.

I knew it was something silly. Thanks for the help anyway!
